New SERVER option "isolation_level" (TODO 2, #250)

ThinkJ001 opened the pull request:
New SERVER option "isolation_level". This gives a chance to use READ COMMITTED instead of SERIALIZABLE for Oracle transactions.

The "isolation_level" value can be serializable(default), read_committed, read_write(not commonly used) and read_only.

laurenz commented:
That looks mostly fine, thanks!

I have two wishes:
1. Let's do away with `read_write`. From my reading of the Oracle documentation, it seems to be just another way to say `READ COMMITTED`. Or do you know more? Anyway, I see no use case for it, and from your comments you seem to feel similar.

2. Can you make `OracleFdwState.isolation_level` an `int` and introduce some `#define`s for it in `oracle_fdw.h`? That avoids messing around with strings (I think that `=` looks nicer than `strcmp(...) = 0`).

  It also avoids repeating string literals in the code - the compiler cannot catch a typo in a string literal.

ThinkJ001 commented:
1. I think you are right, read_write is useless.
2. Sorry, I've not got your idea..., Do you mean use "1" instead of "read_committed",  for example:
CREATE SERVER oracle4 FOREIGN DATA WRAPPER oracle_fdw OPTIONS (dbserver '', isolation_level '1');
something like that?

ThinkJ001 commented:
Hi Laurenz,

1. read_write was removed as you suggested.
2. isolation_level still remain string when CREATE SERVER, for example:
CREATE SERVER oracle4 FOREIGN DATA WRAPPER oracle_fdw OPTIONS (dbserver '', isolation_level 'read_committed');
but it will converted to unsigned int, and passed as unsigned int call between funtions.
Please let me know if that is what you want? Thanks.

laurenz commented:
That looks already very good, you understood what I mean with integers instead of strings (sorry for the terse description).

My remaining complaints are mostly cosmetic:

- The prototype for `getIsolationLevel` should be with the other &ldquo;helper functions&rdquo;, and I try to keep the declatations in the same order as the actual functions, so that I can keep an overview.

- `getIsolationLevel` should not return 0 if the string is wrong. Rather, you should throw an "internal error". It should never occur, but better safe than sorry.

- Could you rename the macro `ORA_TRANS_NEW` to `ORA_TRANS_READ_COMMITTED`?

- In `oracleGetSession` you used spaces instead of tabs. Also, could you use a `switch` statement without a `default` clause there, rather than a cascade of `if`s? That is more readable, and you get a compiler warning if you forget a `case`.

ThinkJ001 commented:
Thank you for your great advice!

1. Now I put the `getIsolationLevel` below `fold_case` (both prototype & implementation)
2. `getIsolaionLevel` may return -1 means invalid option value, and I add a valid value check for option isolation_level in `oracle_fdw_validator`, and will throw an error if provide a invalid value for the option:
CREATE SERVER oracle4 FOREIGN DATA WRAPPER oracle_fdw OPTIONS (dbserver '', isolation_level 'unknow');
ERROR:  invalid value for option "isolation_level"
HINT:  Valid values in this context are: serializable/read_committed/read_only
3. I renamed the macro `ORA_TRANS_NEW` to `ORA_TRANS_READ_COMMITTED`, and `ORA_TRANS_READONLY` to `ORA_TRANS_READ_ONLY`
4. I take your suggestion and changed if s to switch. But if omit default, it will pop a warning when compling:
warning: ‘OCI_TRANS_CODE’ may be used uninitialized in this function
so I assign -1 to OCI_TRANS_CODE when delare the variable, but I doubt if it's appropriate...

laurenz commented:
Great. Thanks again for your efforts.
I'll pick it up from here and get it committed.
I'll be teaching next week, so I don't know when I will find time.

laurenz commented:
For future reference: this implements #250 and #382.
